### PR TITLE
Add `discover_imports` in conf, don't collect imported classes named Test* closes #12749`

### LINKED ISSUE #12749: Don't auto-discover tests in particular folders
This inspired by a discussion: https://github.com/pytest-dev/pytest/discussions/12748

#### What's the problem this feature will solve?

We have a project that's about a different kind of test. We have a lot of classes with "Test" in the name, including some that start with "Test". We would also like to organize some of our pytest tests in classes whose names start with "Test". We distinguish them by directory; all our pytest tests are in `tests`, and all our production code is in `src`. We'd like an easy way to set this up with no distortion to our production code and minimal shenanigans in the test code, while getting no warnings.

Naively one might think that limiting discovery to a particular directory, as with `pytest tests` or `testpaths=["tests"]`. However, if a production file called TestFoo is imported from production code into test code, it will still be discovered, resulting in a warning like:

```
src/pytest_collection_issue/domain.py:6
  src/pytest_collection_issue/domain.py:6: PytestCollectionWarning: cannot collect test class 'Testament' because it has a __init__ constructor (from: tests/test_domain.py)
    class Testament(object):
```

#### Describe the solution you'd like

For me, the best solution is just to by default disable discovery beyond the requested `testpaths`. Trying to run something just because it's included from wherever strikes me as counterintuitive behavior. If somebody needs that behavior, perhaps an option like `discover_imports` would be useful.

The second-best solution would be an option where we can specify paths to definitely not discover things in. E.g.:

```
testpaths = ["tests"]
excludepaths = ["src"]
```

#### Alternative Solutions

There are many ways to solve this, including manually adding a `__test__ = false` to each production class, using a marker mixin like NotAPytestTest to achieve the same end, importing production classes into the tests under different names, doing only local imports, or renaming our test classes. All of these are a little clunky, and require people to remember a ritual under certain circumstances.

The solution I went with is to put

```
for a_class in [i[1] for i in (globals().items()) if inspect.isclass(i[1])]:
    if a_class.__name__.startswith('Test'):
        a_class.__test__ = False
```

after my domain imports and before my test classes. It still seems clunky to me, but only has to happen once per file of tests. I think it would be better still if this could be fixed in one spot, or have it work by default.

#### Additional context

Please see [this GitHub discussion](https://github.com/pytest-dev/pytest/discussions/12748) and [this example repository](https://github.com/wpietri/pytest-collection-issue) that demonstrates the problem.

### PR DESCRIPTION
**Pull request by FreerGit**
closes #12749

- Allows a user to define whether or not pytest should treat imported (but not in testpaths) as test classes.
- Imagine a class is called TestFoo defined in src/ dir, when discover_imports is disabled (default), TestFoo is not treated as a test class.

### CODE REVIEW

File: src/_pytest/main.py
Comment (RonnyPfannschmidt) [inline src/_pytest/main.py:84]: We need to start with a default of true for backwards compatibility 

Alternatively none with a informative warning

File: src/_pytest/python.py
Comment (RonnyPfannschmidt) [inline src/_pytest/python.py:744]: We should ensure this is always a boolean or parsed somewhere else to ensure Separation

**Review by RonnyPfannschmidt — COMMENTED** [review]
Thanks for getting this started 

We have to do some considerations for backwards compatibility and type safety 

Also I'm not yet sure if the selection approach is generally intuitive (it solving your usecase nicely

**Comment by FreerGit** [comment]
@RonnyPfannschmidt Hi! Thanks for reviewing :)

To be clear if I would set True to default this would mean that the base case is that a class like "Testament" in a seperate folder would make pytest error out (see https://github.com/pytest-dev/pytest/issues/12749). Then a user would opt in (set to false) to make sure that the discovery does not happen for those classes. Your suggestion seems more reasonable I think.

I saw some checks failed, which didn't happen locally for some reason. 
I will look over your comments and try and solve this tommorrow!

File: src/_pytest/main.py
Comment (wpietri) [reply src/_pytest/main.py:84]: If it helps, I'd be happy to argue that the current behavior was a substantial surprise to me, so that defaulting to False would count as removing a bug.

Comment (nicoddemus) [reply src/_pytest/main.py:84]: > If it helps, I'd be happy to argue that the current behavior was a substantial surprise to me, so that defaulting to False would count as removing a bug.

I understand where you are coming from and I agree this would be the reasonable default if pytest was being born today... but unfortunately in this case I suspect there are test suites which rely on this behavior, so we really should be conservative here.

File: src/_pytest/main.py
Comment (nicoddemus) [inline src/_pytest/main.py:82]: Is the idea that `discover_imports` is really only active if 'testpaths' is also defined in `pytest.ini`?

File: changelog/12749.feature.rst
Comment (nicoddemus) [inline changelog/12749.feature.rst:1]: Nice, changelog entry.

We also need to add documentation for that confval in `reference.rst`.


I'm not sure about the name too, but at the moment I don't have another suggestion (however we really should strive to find a more appropriate name).

File: src/_pytest/python.py
Comment (nicoddemus) [reply src/_pytest/python.py:744]: I think changing the default value for the option to `False` and changing this to `if self.config.getinit("discover_imports"):` should be enough.

**Comment by nicoddemus** [comment]
I also wonder if we should be using a more general approach, and not only classes: shouldn't we also account for `test_` functions imported from somewhere else (`src`)?

**Comment by nicoddemus** [comment]
Thought about this some more:

I think this should actually be expanded to the general idea/feature of "pytest will only collect objects from modules if they are defined in that module" (by inspecting its `__module__` attribute). So if the user enables that option, classes/functions that appear in a module will only be collected if their `__module__` points to that module.

File: changelog/12749.feature.rst
Comment (FreerGit) [reply changelog/12749.feature.rst:1]: treat_imports_as_tests?
imports_are_tests?
crawl_imports?

Comment (nicoddemus) [reply changelog/12749.feature.rst:1]: As per [my ](https://github.com/pytest-dev/pytest/pull/12810#issuecomment-2348621005), perhaps we should focus on being more "strict" about what we collect from modules.

`collect_?something? = "all"/"local-only"`

But we should wait to see what other maintainers think.

Comment (FreerGit) [reply changelog/12749.feature.rst:1]: Alright, now after a some hours (more than id like to admit) I now have something that works. Both for functions and classes. Let me know how we wanna handle the toggle of the feature. Have a good weekend!

**Comment by nicoddemus** [comment]
Another thing that occurred to me: `__test__` should also be handled by the new option correctly as way to opt-in for something to be considered a test, regardless of where it was imported from.

@pytest-dev/core how do you feel about this feature, should we go ahead with it or are there other concerns to discuss?

**Comment by Zac-HD** [comment]
I'm completely unsurprised that classes or functions imported into a test file are collected, but I'm also obviously atypical of pytest users, and so having a setting for this could make sense.

If we do, I think the setting should be named `collect_imported_tests`; it's not about discovery at all.

> I think this should actually be expanded to the general idea/feature of "pytest will only collect objects from modules if they are defined in that module" (by inspecting its `__module__` attribute). So if the user enables that option, classes/functions that appear in a module will only be collected if their `__module__` points to that module.

Agree with all of this, `__test__ =` definitely takes priority over a global option.  I'd reverse the phrasing a little though; we _should_ collect anything where `__module__` is unset or none even though that doesn't match the current module.

-------------

Finally, I'm concerned that this opens up another avenue for the worst kind of testing bug, where you think you have tests but they can't actually fail (because they're not collected!) and so the whole exercise is silently useless.  I'm unsure whether the benefits are worth this risk, given that there are plenty of other workarounds based on our existing features - including collection patterns, plugins, etc.

**Comment by nicoddemus** [comment]
> I'm unsure whether the benefits are worth this risk, given that there are plenty of other workarounds based on our existing features - including collection patterns, plugins, etc.

I'm not sure about this hiding bugs etc (the behavior seems somewhat more sane for those not used to how pytest works I think), but of course I might be wrong. However I do agree with the general sentiment of "yet another flag" that this brings. In this case however seems reasonable to me (but not that strongly TBH), because if I were designing pytest from the ground up today, without backward compatibility concerns, I would vote that it should behave this way (only collecting classes/functions defined in their own module).

**Comment by wpietri** [comment]
Yeah, as the report of this issue, I found it wildly surprising, and spent a lot of time trying to find a fix. I'd rather have there be a clear option for this, as that's what I first looked for. But failing that, "plenty of other workarounds" was part of the problem for me, as I had to try a lot of stuff and  couldn't find one that let me fix exactly the issue once in a clear and clean way. I am fine with me having to fix it once, but didn't want it to be something every engineer had to learn and remember every time they created a new domain class (or function) with test in the name.

So if something like this doesn't make it in, then I'd at least suggest a clear, official workaround recommendation the docs where people can solve it in one spot permanently for their project.

File: src/_pytest/main.py
Comment (nicoddemus) [inline src/_pytest/main.py:976]: The logic for the variable is inverted: if `collect_imported_tests` is True, we are actually *excluding* imported tests instead of collecting them.

So the default for `collect_imported_tests` should be True for backward compatibility as discussed previously, and the logic here should be inverted:

```suggestion
            if not self.config.getini("collect_imported_tests"):
```

File: src/_pytest/main.py
Comment (nicoddemus) [inline src/_pytest/main.py:991]: My other comment applies here as well.

File: changelog/12749.feature.rst
Comment (nicoddemus) [inline changelog/12749.feature.rst:1]: As per my comment, this description needs an update:

```suggestion
New :confval:`collect_imported_tests`: when enabled (the default) pytest will collect classes/functions in test modules even if they are imported from another file.

Setting this to True will make pytest collect classes/functions from test files only if they are defined in that file (as opposed to imported there).
```

Also, we should describe this option in https://github.com/pytest-dev/pytest/blob/main/doc/en/reference/reference.rst in a new `.. conf-val::` block.

**Review by nicoddemus — CHANGES_REQUESTED** [review]

File: doc/en/reference/reference.rst
Comment (nicoddemus) [inline doc/en/reference/reference.rst:1842]: Can you move this closer to the other options alphabetically?

File: doc/en/reference/reference.rst
Comment (nicoddemus) [inline doc/en/reference/reference.rst:1843]:
```suggestion

   .. versionadded:: 8.4

```

File: doc/en/reference/reference.rst
Comment (nicoddemus) [inline doc/en/reference/reference.rst:1844]:
```suggestion
   Setting this to ``false`` will make pytest collect classes/functions from test
```

File: doc/en/reference/reference.rst
Comment (nicoddemus) [inline doc/en/reference/reference.rst:1847]: RST is fidgety with spaces, this needs to be indented according to the block above otherwise it will not render correctly:

image

(Compare with `tmp_path_retention_count` below).

```suggestion
   .. code-block:: ini
```

File: doc/en/reference/reference.rst
Comment (nicoddemus) [inline doc/en/reference/reference.rst:1850]:
```suggestion
        collect_imported_tests = false
        
   Default: ``true``           
```

File: src/_pytest/main.py
Comment (nicoddemus) [inline src/_pytest/main.py:976]: Previously this code was placed in `_pytest.python`, which I think is the correct place... as it is implemented here, we are calling `pytest_itemcollected` but then discarding the item if the option is set and the item is not defined in its module, which I think is incorrect: `pytest_itemcollected` should only really be called if we did collect the item.

In fact, it is important for this to be tested too: the following collection hooks should not receive any item that was discarded due to `collect_imported_tests`:

* `pytest_itemcollected`
* `pytest_collectreport`
* `pytest_collection_modifyitems`

File: src/_pytest/main.py
Comment (nicoddemus) [inline src/_pytest/main.py:991]: Same comment here, this logic should be moved to `_pytest.python` instead of `main`.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @Bjwebb,

Unfortunately we are not quite there yet.

However feel free to hold on to updating the PR in case other maintainers want to comment further before we move this along.

File: testing/test_collect_imports.py
Comment (nicoddemus) [inline testing/test_collect_imports.py:146]: I think this was meant to be:

```suggestion
        testpaths = "tests"
```

But I don't think `testpaths` is related to the issue anymore?

Comment (FreerGit) [reply testing/test_collect_imports.py:146]: I ment delete that commented line but forgot. You're right `testpaths` is not related and I wanted to make a test to ensure that.

Comment (nicoddemus) [reply testing/test_collect_imports.py:146]: Oh in that case make sure to add a comment/docstring mentioning that if you want to keep the test.

File: src/_pytest/main.py
Comment (FreerGit) [reply src/_pytest/main.py:976]: Hi! Thought I would pick up this issue again after some wait. Is there any examples on how I could test these hooks? Im unsure on how to extract that information from the hooks. Im assuming this is problematic because extensions to pytest could be listening to these hooks?

Comment (nicoddemus) [reply src/_pytest/main.py:976]: > Is there any examples on how I could test these hooks? Im unsure on how to extract that information from the hooks.

Usually we write a `conftest.py` file in the tests that implement these hooks, and then set/update some global variable when they are called.

> Im assuming this is problematic because extensions to pytest could be listening to these hooks?

Exactly.

File: testing/test_collect_imports.py
Comment (FreerGit) [inline testing/test_collect_imports.py:228]: Im not completely sure what  is the correct behaviour yet, I'll try and figure it out. WIP and all.

File: testing/test_collect_imports.py
Comment (FreerGit) [inline testing/test_collect_imports.py:328]: At this point, I need some guidance.

The problem now is that while the CollectReport's for classes and functions are correct, the directory gets a report (even though no classes/functions were collected from it). What should be the expected behaviour in this case? Since the directory will be collected before any of the classes/functions I would need to filter it out later on somehow. Not sure what the correct approach is here. 

I do feel like collecting the report for the directory (even though it's not "used") makes sense but I would need a maintainer to check.

Comment (nicoddemus) [reply testing/test_collect_imports.py:328]: > The problem now is that while the CollectReport's for classes and functions are correct, the directory gets a report (even though no classes/functions were collected from it). What should be the expected behaviour in this case?

This is fine -- currently on `main` if we implement `pytest_collectreport` and run pytest on a bunch of directories without any tests, we get collection reports for all directories, even though no teste items were collected. 

My previous concern was that we were generating collection reports for items that were in fact being discarded later, and that would definitely be  a problem -- but collection reports for directories it is expected that some of them might turn out to not collect anything.

**Comment by nicoddemus** [comment]
Thanks @FreerGit for the patience and sorry for the delay!

I changed your original implementation a bit: we only need to prevent the collection of classes/functions at the module level, because at that point they will not even appear as a `Class` later for their methods to be collected. This simplifies the implementation and tests significantly.

Also improved the docs a bit providing an example for easier understanding.

Thanks again!

@RonnyPfannschmidt could you review it too? Thanks.

**Review by nicoddemus — APPROVED** [review]

**Comment by nicoddemus** [comment]
Ping @Zac-HD in case he wants to review too. 👍

**Review by Zac-HD — APPROVED** [review]
Looks great - thanks everyone!

**Comment by nicoddemus** [comment]
@Zac-HD please remember to squash next time. 😢 

Also I would have liked for @RonnyPfannschmidt's review before merging.

**Comment by Zac-HD** [comment]
Gah, sorry!  I was rushing out and should have left it for when I had more time 😓

**Comment by nicoddemus** [comment]
No worries, it happens. :+1:

**Comment by FreerGit** [comment]
Great to see this PR getting merged! Thanks for all the help and appreciate the cleanup @nicoddemus , admittedly the PR required more understanding of the internals than I first thought...

Anyhow, great to see it move along :)

**Pull request merged**
